File: frontend/src/types/match.ts
import {IPerk, ISpell} from "./store";

export interface IMatchHistory {
    accountId: number;
    games: {
        gameBeginDate: string;
        gameCount: number;
        gameEndDate: string;
        gameIndexBegin: number;
        gameIndexEnd: number;
        games: IGame[];
    };
    platformId: string;
}

export interface IMatchData {
    gameCreation: number;
    gameCreationDate: string;
    gameDuration: number;
    gameId: number;
    gameMode: string;
    gameType: string;
    gameVersion: string;
    mapId: number;
    participantIdentities: IParticipantIdentity[];
    participants: IParticipant[];
    platformId: string;
    queueId: number;
    seasonId: number;
    teams: ITeam[];
}

export interface IDisplayMatch {
    poz: number; // pozitia in array a player ului cautat
    queue: string; // queue ul fiecarui meci
    gameDuration: string;
    gameTimeAgo: string | null;
    kp: string;
    spellD: ISpell; // pe pozitia 0 am spellu ul in sine, pe poz 1 am descrierea, pe poz 2 am numele
    spellF: ISpell;
    runePrimary: IPerk; // pe pozitia 0 am runa in sine, pe pozitia 1 am descrierea, pe pozitia 2 am numele
    runeSecondaryStyle: number;
    toggleAdvancedDetails: boolean;
    realChampsNames: string[];  // un array cu toate numele reale ale campionilor, pt ca unii nu corespund ex: MonkeyKing - Wukong
}

export interface IMatchStatistic {
    mainPlayerPoz: number[]; // array with all positions of main player in all matches
    kills: number;
    assists: number;
    deaths: number;
    wins: number;
    defeats: number;

    gold: number;
    vision_score: number;
    firstBloodTimes: number;
    cs: number;
    control_wards: number;

    lane: string[]; // vector cu toate lane-urile
    preferablyLane: string | null; // cel mai jucat lane
    preferablyLaneGames: number; // nr meciurilor jucate pe cel mai jucat lane
    champs: number[]; // vector cu toți campionii
    preferablyChamp1: number | null; // cel mai jucat campion
    preferablyChamp2: number | null; // al doilea cel mai jucat campion

    preferablyChamp1_games: number;
    preferablyChamp2_games: number;

    kills_preferablyLane: number; // statistici pe cel mai jucat lane
    deaths_preferablyLane: number;
    assists_preferablyLane: number;
    wins_preferablyLane: number;
    defeats_preferablyLane: number;

    kills_preferablyChamps: [number, number]; // statistici pentru cei mai jucati campioni [0: cel mai jucat champ, 1: al 2-lea cel mai jucat campion]
    deaths_preferablyChamps: [number, number];
    assists_preferablyChamps: [number, number];
    wins_preferablyChamps: [number, number];
    defeats_preferablyChamps: [number, number];
}

interface IParticipantIdentity {
    participantId: number;
    player: {
        accountId: number;
        currentAccountId: number;
        currentPlatformId: string;
        matchHistoryUri: string;
        platformId: string;
        profileIcon: number;
        puuid:string,
        summonerId: number;
        summonerName: string;
    };
}

interface IParticipant {
    championId: number;
    highestAchievedSeasonTier: string;
    participantId: number;
    spell1Id: number;
    spell2Id: number;
    stats: IParticipantStats;
    teamId: number;
    timeline: IParticipantTimeline;
}

interface IParticipantStats {
    assists: number;
    causedEarlySurrender: boolean;
    champLevel: number;
    combatPlayerScore: number;
    damageDealtToObjectives: number;
    damageDealtToTurrets: number;
    damageSelfMitigated: number;
    deaths: number;
    doubleKills: number;
    earlySurrenderAccomplice: boolean;
    firstBloodAssist: boolean;
    firstBloodKill: boolean;
    firstInhibitorAssist: boolean;
    firstInhibitorKill: boolean;
    firstTowerAssist: boolean;
    firstTowerKill: boolean;
    gameEndedInEarlySurrender: boolean;
    gameEndedInSurrender: boolean;
    goldEarned: number;
    goldSpent: number;
    inhibitorKills: number;
    item0: number;
    item1: number;
    item2: number;
    item3: number;
    item4: number;
    item5: number;
    item6: number;
    killingSprees: number;
    kills: number;
    largestCriticalStrike: number;
    largestKillingSpree: number;
    largestMultiKill: number;
    longestTimeSpentLiving: number;
    magicDamageDealt: number;
    magicDamageDealtToChampions: number;
    magicalDamageTaken: number;
    neutralMinionsKilled: number;
    neutralMinionsKilledEnemyJungle: number;
    neutralMinionsKilledTeamJungle: number;
    objectivePlayerScore: number;
    participantId: number;
    pentaKills: number;
    perk0: number;
    perk0Var1: number;
    perk0Var2: number;
    perk0Var3: number;
    perk1: number;
    perk1Var1: number;
    perk1Var2: number;
    perk1Var3: number;
    perk2: number;
    perk2Var1: number;
    perk2Var2: number;
    perk2Var3: number;
    perk3: number;
    perk3Var1: number;
    perk3Var2: number;
    perk3Var3: number;
    perk4: number;
    perk4Var1: number;
    perk4Var2: number;
    perk4Var3: number;
    perk5: number;
    perk5Var1: number;
    perk5Var2: number;
    perk5Var3: number;
    perkPrimaryStyle: number;
    perkSubStyle: number;
    physicalDamageDealt: number;
    physicalDamageDealtToChampions: number;
    physicalDamageTaken: number;
    playerScore0: number;
    playerScore1: number;
    playerScore2: number;
    playerScore3: number;
    playerScore4: number;
    playerScore5: number;
    playerScore6: number;
    playerScore7: number;
    playerScore8: number;
    playerScore9: number;
    quadraKills: number;
    sightWardsBoughtInGame: number;
    teamEarlySurrendered: boolean;
    timeCCingOthers: number;
    totalDamageDealt: number;
    totalDamageDealtToChampions: number;
    totalDamageTaken: number;
    totalHeal: number;
    totalMinionsKilled: number;
    totalPlayerScore: number;
    totalScoreRank: number;
    totalTimeCrowdControlDealt: number;
    totalUnitsHealed: number;
    tripleKills: number;
    trueDamageDealt: number;
    trueDamageDealtToChampions: number;
    trueDamageTaken: number;
    turretKills: number;
    unrealKills: number;
    visionScore: number;
    visionWardsBoughtInGame: number;
    wardsKilled: number;
    wardsPlaced: number;
    win: boolean;
}

interface ITeam {
    bans: {
        championId: number;
        pickTurn: number;
    }[];
    baronKills: number;
    dominionVictoryScore: number;
    dragonKills: number;
    firstBaron: boolean;
    firstBlood: boolean;
    firstDargon: boolean;
    firstInhibitor: boolean;
    firstTower: boolean;
    inhibitorKills: number;
    riftHeraldKills: number;
    teamId: number;
    towerKills: number;
    vilemawKills: number;
    win: string;
}

interface IGame {
    gameCreation: number;
    gameCreationDate: string;
    gameDuration: number;
    gameId: number;
    gameMode: string;
    gameType: string;
    gameVersion: string;
    mapId: number;
    participantIdentities: IParticipantIdentity[];
    participants: IParticipant[];
    platformId: string;
    queueId: number;
    seasonId: number;
    teams: ITeam[];
}

interface IParticipantTimeline {
    creepsPerMinDeltas: {
        [key: string]: number;
    };
    csDiffPerMinDeltas: {
        [key: string]: number;
    };
    damageTakenDiffPerMinDeltas: {
        [key: string]: number;
    };
    damageTakenPerMinDeltas: {
        [key: string]: number;
    };
    goldPerMinDeltas: {
        [key: string]: number;
    };
    lane: string;
    participantId: number;
    xpDiffPerMinDeltas: {
        [key: string]: number;
    };
    xpPerMinDeltas: {
        [key: string]: number;
    };
}

File: backend/model/match.go
package model

// MatchData 查询出的JSON一场比赛细节数据
type MatchData struct {
	GameCreation          int                   `json:"gameCreation"`
	GameCreationDate      string                `json:"gameCreationDate"`
	GameDuration          int                   `json:"gameDuration"`
	GameId                int                   `json:"gameId"`
	GameMode              string                `json:"gameMode"`
	GameType              string                `json:"gameType"`
	GameVersion           string                `json:"gameVersion"`
	MapId                 int                   `json:"mapId"`
	ParticipantIdentities []ParticipantIdentity `json:"participantIdentities"`
	Participants          []Participant         `json:"participants"`
	PlatformId            string                `json:"platformId"`
	QueueId               int                   `json:"queueId"`
	SeasonId              int                   `json:"seasonId"`
	Teams                 []Team                `json:"teams"`
}

type Game struct {
	GameCreation          int                   `json:"gameCreation"`
	GameCreationDate      string                `json:"gameCreationDate"`
	GameDuration          int                   `json:"gameDuration"`
	GameId                int                   `json:"gameId"`
	GameMode              string                `json:"gameMode"`
	GameType              string                `json:"gameType"`
	GameVersion           string                `json:"gameVersion"`
	MapId                 int                   `json:"mapId"`
	ParticipantIdentities []ParticipantIdentity `json:"participantIdentities"`
	Participants          []Participant         `json:"participants"`
	PlatformId            string                `json:"platformId"`
	QueueId               int                   `json:"queueId"`
	SeasonId              int                   `json:"seasonId"`
	Teams                 []Team                `json:"teams"`
}

type Games struct {
	GameBeginDate  string `json:"gameBeginDate"`
	GameCount      int    `json:"gameCount"`
	GameEndDate    string `json:"gameEndDate"`
	GameIndexBegin int    `json:"gameIndexBegin"`
	GameIndexEnd   int    `json:"gameIndexEnd"`
	Games          []Game `json:"games"`
}

type MatchHistory struct {
	AccountId  int    `json:"accountId"`
	Games      Games  `json:"games"`
	PlatformId string `json:"platformId"`
}

type Participant struct {
	ChampionId                int      `json:"championId"`
	HighestAchievedSeasonTier string   `json:"highestAchievedSeasonTier"`
	ParticipantId             int      `json:"participantId"`
	Spell1Id                  int      `json:"spell1Id"`
	Spell2Id                  int      `json:"spell2Id"`
	Stats                     Stats    `json:"stats"`
	TeamId                    int      `json:"teamId"`
	Timeline                  Timeline `json:"timeline"`
}

type Timeline struct {
	CreepsPerMinDeltas          map[string]float64 `json:"creepsPerMinDeltas"`
	CsDiffPerMinDeltas          map[string]float64 `json:"csDiffPerMinDeltas"`
	DamageTakenDiffPerMinDeltas map[string]float64 `json:"damageTakenDiffPerMinDeltas"`
	DamageTakenPerMinDeltas     map[string]float64 `json:"damageTakenPerMinDeltas"`
	GoldPerMinDeltas            map[string]float64 `json:"goldPerMinDeltas"`
	Lane                        string             `json:"lane"`
	Role                        string             `json:"role"`
	ParticipantId               int                `json:"participantId"`
	XpDiffPerMinDeltas          map[string]float64 `json:"xpDiffPerMinDeltas"`
	XpPerMinDeltas              map[string]float64 `json:"xpPerMinDeltas"`
}

type Stats struct {
	Assists                         int  `json:"assists"`
	CausedEarlySurrender            bool `json:"causedEarlySurrender"`
	ChampLevel                      int  `json:"champLevel"`
	CombatPlayerScore               int  `json:"combatPlayerScore"`
	DamageDealtToObjectives         int  `json:"damageDealtToObjectives"`
	DamageDealtToTurrets            int  `json:"damageDealtToTurrets"`
	DamageSelfMitigated             int  `json:"damageSelfMitigated"`
	Deaths                          int  `json:"deaths"`
	DoubleKills                     int  `json:"doubleKills"`
	EarlySurrenderAccomplice        bool `json:"earlySurrenderAccomplice"`
	FirstBloodAssist                bool `json:"firstBloodAssist"`
	FirstBloodKill                  bool `json:"firstBloodKill"`
	FirstInhibitorAssist            bool `json:"firstInhibitorAssist"`
	FirstInhibitorKill              bool `json:"firstInhibitorKill"`
	FirstTowerAssist                bool `json:"firstTowerAssist"`
	FirstTowerKill                  bool `json:"firstTowerKill"`
	GameEndedInEarlySurrender       bool `json:"gameEndedInEarlySurrender"`
	GameEndedInSurrender            bool `json:"gameEndedInSurrender"`
	GoldEarned                      int  `json:"goldEarned"`
	GoldSpent                       int  `json:"goldSpent"`
	InhibitorKills                  int  `json:"inhibitorKills"`
	Item0                           int  `json:"item0"`
	Item1                           int  `json:"item1"`
	Item2                           int  `json:"item2"`
	Item3                           int  `json:"item3"`
	Item4                           int  `json:"item4"`
	Item5                           int  `json:"item5"`
	Item6                           int  `json:"item6"`
	KillingSprees                   int  `json:"killingSprees"`
	Kills                           int  `json:"kills"`
	LargestCriticalStrike           int  `json:"largestCriticalStrike"`
	LargestKillingSpree             int  `json:"largestKillingSpree"`
	LargestMultiKill                int  `json:"largestMultiKill"`
	LongestTimeSpentLiving          int  `json:"longestTimeSpentLiving"`
	MagicDamageDealt                int  `json:"magicDamageDealt"`
	MagicDamageDealtToChampions     int  `json:"magicDamageDealtToChampions"`
	MagicalDamageTaken              int  `json:"magicalDamageTaken"`
	NeutralMinionsKilled            int  `json:"neutralMinionsKilled"`
	NeutralMinionsKilledEnemyJungle int  `json:"neutralMinionsKilledEnemyJungle"`
	NeutralMinionsKilledTeamJungle  int  `json:"neutralMinionsKilledTeamJungle"`
	ObjectivePlayerScore            int  `json:"objectivePlayerScore"`
	ParticipantId                   int  `json:"participantId"`
	PentaKills                      int  `json:"pentaKills"`
	Perk0                           int  `json:"perk0"`
	Perk0Var1                       int  `json:"perk0Var1"`
	Perk0Var2                       int  `json:"perk0Var2"`
	Perk0Var3                       int  `json:"perk0Var3"`
	Perk1                           int  `json:"perk1"`
	Perk1Var1                       int  `json:"perk1Var1"`
	Perk1Var2                       int  `json:"perk1Var2"`
	Perk1Var3                       int  `json:"perk1Var3"`
	Perk2                           int  `json:"perk2"`
	Perk2Var1                       int  `json:"perk2Var1"`
	Perk2Var2                       int  `json:"perk2Var2"`
	Perk2Var3                       int  `json:"perk2Var3"`
	Perk3                           int  `json:"perk3"`
	Perk3Var1                       int  `json:"perk3Var1"`
	Perk3Var2                       int  `json:"perk3Var2"`
	Perk3Var3                       int  `json:"perk3Var3"`
	Perk4                           int  `json:"perk4"`
	Perk4Var1                       int  `json:"perk4Var1"`
	Perk4Var2                       int  `json:"perk4Var2"`
	Perk4Var3                       int  `json:"perk4Var3"`
	Perk5                           int  `json:"perk5"`
	Perk5Var1                       int  `json:"perk5Var1"`
	Perk5Var2                       int  `json:"perk5Var2"`
	Perk5Var3                       int  `json:"perk5Var3"`
	PerkPrimaryStyle                int  `json:"perkPrimaryStyle"`
	PerkSubStyle                    int  `json:"perkSubStyle"`
	PhysicalDamageDealt             int  `json:"physicalDamageDealt"`
	PhysicalDamageDealtToChampions  int  `json:"physicalDamageDealtToChampions"`
	PhysicalDamageTaken             int  `json:"physicalDamageTaken"`
	PlayerScore0                    int  `json:"playerScore0"`
	PlayerScore1                    int  `json:"playerScore1"`
	PlayerScore2                    int  `json:"playerScore2"`
	PlayerScore3                    int  `json:"playerScore3"`
	PlayerScore4                    int  `json:"playerScore4"`
	PlayerScore5                    int  `json:"playerScore5"`
	PlayerScore6                    int  `json:"playerScore6"`
	PlayerScore7                    int  `json:"playerScore7"`
	PlayerScore8                    int  `json:"playerScore8"`
	PlayerScore9                    int  `json:"playerScore9"`
	QuadraKills                     int  `json:"quadraKills"`
	SightWardsBoughtInGame          int  `json:"sightWardsBoughtInGame"`
	TeamEarlySurrendered            bool `json:"teamEarlySurrendered"`
	TimeCCingOthers                 int  `json:"timeCCingOthers"`
	TotalDamageDealt                int  `json:"totalDamageDealt"`
	TotalDamageDealtToChampions     int  `json:"totalDamageDealtToChampions"`
	TotalDamageTaken                int  `json:"totalDamageTaken"`
	TotalHeal                       int  `json:"totalHeal"`
	TotalMinionsKilled              int  `json:"totalMinionsKilled"`
	TotalPlayerScore                int  `json:"totalPlayerScore"`
	TotalScoreRank                  int  `json:"totalScoreRank"`
	TotalTimeCrowdControlDealt      int  `json:"totalTimeCrowdControlDealt"`
	TotalUnitsHealed                int  `json:"totalUnitsHealed"`
	TripleKills                     int  `json:"tripleKills"`
	TrueDamageDealt                 int  `json:"trueDamageDealt"`
	TrueDamageDealtToChampions      int  `json:"trueDamageDealtToChampions"`
	TrueDamageTaken                 int  `json:"trueDamageTaken"`
	TurretKills                     int  `json:"turretKills"`
	UnrealKills                     int  `json:"unrealKills"`
	VisionScore                     int  `json:"visionScore"`
	VisionWardsBoughtInGame         int  `json:"visionWardsBoughtInGame"`
	WardsKilled                     int  `json:"wardsKilled"`
	WardsPlaced                     int  `json:"wardsPlaced"`
	Win                             bool `json:"win"`
}

type ParticipantIdentity struct {
	ParticipantId int `json:"participantId"`
	Player        struct {
		AccountId         int    `json:"accountId"`
		CurrentAccountId  int    `json:"currentAccountId"`
		CurrentPlatformId string `json:"currentPlatformId"`
		MatchHistoryUri   string `json:"matchHistoryUri"`
		PlatformId        string `json:"platformId"`
		ProfileIcon       int    `json:"profileIcon"`
		Puuid             string `json:"puuid"`
		SummonerId        int    `json:"summonerId"`
		SummonerName      string `json:"summonerName"`
	} `json:"player"`
}

type Team struct {
	Bans                 []Ban  `json:"bans"`
	BaronKills           int    `json:"baronKills"`
	DominionVictoryScore int    `json:"dominionVictoryScore"`
	DragonKills          int    `json:"dragonKills"`
	FirstBaron           bool   `json:"firstBaron"`
	FirstBlood           bool   `json:"firstBlood"`
	FirstDragon          bool   `json:"firstDragon"`
	FirstInhibitor       bool   `json:"firstInhibitor"`
	FirstTower           bool   `json:"firstTower"`
	InhibitorKills       int    `json:"inhibitorKills"`
	RiftHeraldKills      int    `json:"riftHeraldKills"`
	TeamId               int    `json:"teamId"`
	TowerKills           int    `json:"towerKills"`
	VilemawKills         int    `json:"vilemawKills"`
	Win                  string `json:"win"`
}

type Ban struct {
	ChampionId int `json:"championId"`
	PickTurn   int `json:"pickTurn"`
}
